File: 032-Longest-Valid-Parentheses/Solution 2.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        // 1. setup a mark array
        // 2. mark all valid parenteses
        // 3. find the longest sub-array in array which's values are true
        int len = s.size();
        // 1
        bool *marked = new bool[len];
        memset(marked, 0, len);
        stack<int> visited;
        
        // 2
        for(int i=0; i < len; i++){
            if(s[i] == '(')
                visited.push(i);
            else if(s[i] == ')' && !visited.empty()){
                marked[i] = true;
                marked[visited.top()] = true;
                visited.pop();
            }
        }
        // 3
        int curr_len = 0, max_len = 0;
        for(int i = 0; i < len; i++){
            if(marked[i] == true)
                curr_len++;
            else{
                max_len = max(max_len, curr_len);
                curr_len = 0;
            }
        }
        max_len = max(max_len, curr_len);
        return max_len;
    }
};

File: Solutions/005-Longest-Palindromic-Substring/Solution2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        """
        dp[i][j] = 1 <-> dp[i+1]dp[j-1] == 1 and s[i] == s[j]
        """
        n = len(s)
        if n <= 1:
            return s
        
        dp = [[False for _ in range(n)] for _ in range(n)]
        max_len, max_s = 1, s[0]
        for j in range(n):
            dp[j][j] = True
            for i in range(j-1, -1, -1):
                if i+1 <= j-1 and dp[i+1][j-1] and s[i] == s[j]:
                    dp[i][j] = True
                elif i+1 > j-1 and s[i] == s[j]:
                    dp[i][j] = True
                
                if dp[i][j] and  max_len < j-i+1:
                    max_len = j-i+1
                    max_s = s[i:j+1]
                
        
        return max_s
